File: CodingSamples/DotNet/Fundamentals/Generics/EnumerableTest/DemoApp1/Common/SimpleStack.cs
namespace DemoApp.Common;

//open generic type 
public class SimpleStack<E> : IStackReader<E>, IStackWriter<E>
{
    class Node
    {
        internal E element;
        internal Node below;

        internal Node Skip(int steps)
        {
            Node n = this;
            for(int i = 0; i < steps; ++i)
                n = n.below;
            return n;
        }
    }

    private Node top;

    public void Push(E item)
    {
        top = new Node { element = item, below = top };
    }

    public E Pop()
    {
        Node temp = top;
        top = top.below;
        return temp.element;
    }

    public bool Empty()
    {
        return top is null;
    }

    //indexer - a parameterized property for getting/setting
    //a value within this instance indexed by the parameters
    public E this[int index]
    {
        get{ return top.Skip(index).element; }
        set{ top.Skip(index).element = value; }
    }

    //to support standard iteration pattern, a class must include a
    //public definition for GetEnumerator() method whose return type
    //must expose MoveNext() method (to determine whether an element
    //is available) and a gettable Current property (to obtain that element)
    public Iterator GetEnumerator()
    {
        return new Iterator(this);
    }

    public struct Iterator(SimpleStack<E> source)
    {
        private Node next = source.top;

        public E Current { get; private set; }

        public bool MoveNext()
        {
            if(next != null)
            {
                Current = next.element;
                next = next.below;
                return true;
            }
            return false;
        }

    }
}

File: CodingSamples/Java/Fundamentals/Generics/IterableTest/demo1/SimpleStack.java
import java.util.Iterator;

//in order to support standard iteration in java.
//a class must implement java.lang.Iteratble<T> interface
class SimpleStack<E> implements Iterable<E> {
    
    //inner member class
    class Node {

        E element;
        Node below;

        Node(E value) {
            element = value;
            below = top;
        }
    }

    private Node top;

    public void push(E item) {
        top = new Node(item);
    }

    public E pop() {
        Node temp = top;
        top = top.below;
        return temp.element;
    }

    public boolean empty() {
        return top == null;
    }

    public Iterator<E> iterator() {
        //returning a new instance of an inner local anonymous class
        //that implements Iterator<E>
        return new Iterator<E>(){

            private Node current = top;

            public boolean hasNext() {
                return current != null;
            }

            public E next() {
                E value = current.element;
                current = current.below;
                return value;
            }
        };
    }
}
